File: farmaciaSpring/src/main/java/com/example/sistemafarmacia/service/ProductoService.java
package com.example.sistemafarmacia.service;

import com.example.sistemafarmacia.model.Producto;
import java.util.List;

public interface ProductoService {
    // Búsquedas
    List<Producto> buscarPorNombre(String nombre);
    List<Producto> buscarPorCategoria(String categoria);
    
    // CRUD
    List<Producto> getAllProductos();
    Producto getProductoById(int id);
    Producto addProducto(Producto producto);
    Producto updateProducto(int id, Producto producto);
    void deleteProducto(int id);
}

File: Farmacia-app/src/app/productos/producto.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Producto } from '../models/producto.model';

@Injectable({
  providedIn: 'root'
})
export class ProductoService {
  private baseUrl = '/api/producto';

  constructor(private http: HttpClient) { }

  getProductos(): Observable<Producto[]> {
    return this.http.get<Producto[]>(this.baseUrl);
  }

  getProducto(id: number): Observable<Producto> {
    return this.http.get<Producto>(`${this.baseUrl}/${id}`);
  }

  createProducto(producto: Producto): Observable<Producto> {
    return this.http.post<Producto>(this.baseUrl, producto);
  }

  updateProducto(id: number, producto: Producto): Observable<Producto> {
    return this.http.put<Producto>(`${this.baseUrl}/${id}`, producto);
  }

  deleteProducto(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  buscarPorNombre(nombre: string): Observable<Producto[]> {
    return this.http.get<Producto[]>(`${this.baseUrl}/buscar/nombre?nombre=${nombre}`);
  }

  buscarPorCategoria(categoria: string): Observable<Producto[]> {
    return this.http.get<Producto[]>(`${this.baseUrl}/buscar/categoria?categoria=${categoria}`);
  }
}
